File: internal/video/editor_audio_ducking.go
package video

import (
	"fmt"
	"math"
	"sort"
	"strings"
)

// Keep this contract identical to editorAudioDucking.ts: fades are inside
// merged voice-over windows, shortened to half the window for short segments.
const originalAudioDuckingGain = 0.25
const audioDuckingFadeSeconds = 0.2

type audioDuckingWindow struct{ start, end float64 }

func audioDuckingWindows(timeline editTimeline) []audioDuckingWindow {
	var windows []audioDuckingWindow
	for _, segment := range renderAudioSegments(timeline.Clips, timeline.AudioSegments) {
		if audioTrackID(segment) != "voiceover-1" || segment.Muted {
			continue
		}
		speed, err := effectiveRenderAudioSpeed(segment, timeline.Clips)
		if err != nil { // Render validation rejects invalid geometry before this stage.
			continue
		}
		end := segment.TimelineStart + (segment.SourceEnd-segment.SourceStart)/speed
		if end > segment.TimelineStart {
			windows = append(windows, audioDuckingWindow{segment.TimelineStart, end})
		}
	}
	sort.SliceStable(windows, func(i, j int) bool { return windows[i].start < windows[j].start })
	var merged []audioDuckingWindow
	for _, window := range windows {
		if n := len(merged); n > 0 && window.start <= merged[n-1].end {
			merged[n-1].end = math.Max(merged[n-1].end, window.end)
		} else {
			merged = append(merged, window)
		}
	}
	return merged
}

func audioDuckingExpression(windows []audioDuckingWindow) string {
	expression := "1"
	for i := len(windows) - 1; i >= 0; i-- {
		w := windows[i]
		fade := math.Min(audioDuckingFadeSeconds, (w.end-w.start)/2)
		expression = fmt.Sprintf("if(gte(t,%.9f)*lt(t,%.9f),1-%.9f*min(1,min((t-%.9f)/%.9f,(%.9f-t)/%.9f)),%s)",
			w.start, w.end, 1-originalAudioDuckingGain, w.start, fade, w.end, fade, expression)
	}
	return expression
}

// The established builder has already applied source offsets, speed, volume,
// mute, gaps and exact duration to each stereo/48-kHz track. Only multiply the
// original track before amix. aeval evaluates per sample, not per audio frame.
// All video/overlay arguments and the voice-over track remain untouched.
func applyTimelineDucking(args []string, timeline editTimeline) []string {
	if timeline.DuckOriginalAudio == nil || !*timeline.DuckOriginalAudio {
		return args
	}
	windows := audioDuckingWindows(timeline)
	if len(windows) == 0 {
		return args
	}
	const mix = "[mix0][mix1]amix="
	for i := 0; i+1 < len(args); i++ {
		if args[i] != "-filter_complex" || !strings.Contains(args[i+1], mix) {
			continue // Original-only / voice-over-only graphs need no ducking.
		}
		gain := audioDuckingExpression(windows)
		// Two expressions alone can negotiate an unnamed "2 channels" layout
		// through concat/amix. AAC requires the explicit stereo speaker layout.
		filter := fmt.Sprintf("[mix0]aeval=exprs='val(0)*(%s)|val(1)*(%s)':channel_layout=stereo,aformat=sample_fmts=fltp:sample_rates=48000:channel_layouts=stereo[duckedOriginal];[duckedOriginal][mix1]amix=", gain, gain)
		result := append([]string(nil), args...)
		result[i+1] = strings.Replace(args[i+1], mix, filter, 1)
		return result
	}
	return args
}

File: web/src/components/editorAudioDucking.ts
import { audioSegmentTimelineDuration, audioTrackId, type EditorAudioSegment } from "./editorAudioGeometry";
import type { EditorClip } from "./editorClipTime";

export const ORIGINAL_AUDIO_DUCKING_GAIN = 0.25;
export const AUDIO_DUCKING_FADE_SECONDS = 0.2;

// Pure preparation only: not connected to preview, persistence or render.
// Returns a multiplier for the original segment's volume, not a replacement.
// Call with the editor's validated segments and current clip geometry.
export function originalAudioDuckingGain(
  segments: readonly EditorAudioSegment[],
  clips: EditorClip[],
  timelineTime: number,
): number {
  if (!Number.isFinite(timelineTime)) return 1;

  const windows = segments
    .filter(segment => audioTrackId(segment) === "voiceover-1" && segment.muted !== true)
    .map(segment => ({
      start: segment.timelineStart,
      end: segment.timelineStart + audioSegmentTimelineDuration(segment, clips),
    }))
    .filter(window => window.end > window.start)
    .sort((a, b) => a.start - b.start);

  // Merge touching/overlapping windows before fading: splits must not cause a
  // volume bump, and simultaneous voice-overs must never multiply attenuation.
  const merged: typeof windows = [];
  for (const window of windows) {
    const previous = merged.at(-1);
    if (previous && window.start <= previous.end) {
      previous.end = Math.max(previous.end, window.end);
    } else {
      merged.push({ ...window });
    }
  }

  const active = merged.find(window => timelineTime >= window.start && timelineTime < window.end);
  if (!active) return 1;

  // Both fades stay inside the active range; before/after it gain is exactly 1.
  // Short windows shorten both fades equally and reach the target at midpoint.
  const fade = Math.min(AUDIO_DUCKING_FADE_SECONDS, (active.end - active.start) / 2);
  const depth = Math.min(1, (timelineTime - active.start) / fade, (active.end - timelineTime) / fade);
  return 1 - (1 - ORIGINAL_AUDIO_DUCKING_GAIN) * depth;
}
